### ISSUE TITLE
[Benchmark][Tracking] Make performance baselines correct, competitive, and protocol-consistent

### ISSUE THREAD

**Issue by superAngGao** [tile-ai/TileOPs#1880]
## Motivation

The nightly dashboard currently treats several different things as if they
were interchangeable performance baselines:

- direct framework or production-library implementations;
- benchmark-local decompositions with extra casts or kernels;
- high-precision correctness references;
- external baselines that do not follow the same per-iteration input policy;
- tags that describe neither the actual provider nor whether the result is a
  performance comparison.

This can produce valid correctness checks but misleading speedup conclusions.
The 2026-08-10 H200 report exposed concrete instances in reductions,
normalization, FP8 GEMM, attention, FLA/Mamba adapters, and FP8 GQA.

Audited commit: `3836e9b`  
Related work: #1806, #1837, #1838

## Problem classes

1. **Baseline implementation:** the timed callable performs unnecessary FP32
   materialization, decomposes an available native API, or includes unrelated
   setup/state-copy work.
2. **Weak or missing competitive baseline:** a high-precision or decomposed
   reference may be the only implementation available. It remains a valid
   baseline, but a stronger contract-equivalent framework/library result should
   be added where one exists.
3. **Tag correctness:** some tags misidentify the timed implementation, and
   reference tags are not consistently separated from competitive baselines.
4. **Per-op protocol deviations:** an adapter bypasses the shifting input pool
   or overrides the suite sampling protocol.

The comments below contain the operator-level audit and proposed correction for
each class.

## Baseline contract

- Keep correctness references as valid baselines. When no stronger compatible
  implementation exists, a clearly named reference may remain the only
  comparison and may still appear in speedup/regression reporting.
- Add a native framework API or production kernel alongside the reference
  wherever one exists. Use an exact same-dtype compiled graph when useful, but
  do not make it a requirement for keeping a benchmark reportable.
- Keep the existing `tag` schema. Use implementation tags such as
  `flashinfer`, `fa3`, `torch`, and `torch-scaled-mm` for competitive
  baselines, and explicit `-ref` tags such as `torch-ref`, `torch-fp32-ref`,
  and `torch-autograd-ref` for correctness references.
- Performance baselines must use the same CUPTI sampling, cache preparation,
  and effective input policy as TileOps. Library planning/JIT may remain outside
  timing, but per-invocation data preparation required by the contract may not.
- Keep the existing report-selection behavior. Show every recorded baseline;
  an accurate tag tells readers whether a comparison is against FlashInfer,
  native PyTorch, or a benchmark-local reference.

## Scope

This umbrella tracks benchmark construction, performance-baseline coverage,
provenance, and protocol consistency. It does **not** track kernel optimization,
adjudicate whether a large direct-API speedup is genuine, or turn every slower
TileOps result into a performance issue.

## Acceptance criteria

- [ ] Confirmed cast/decomposition/setup inflation is removed from performance
      baselines.
- [ ] Reference-only operators listed in the comments gain a stronger concrete
      implementation where one is compatible; otherwise the accurately tagged
      reference remains available.
- [ ] Tags accurately name the callable that was timed, without adding manifest
      fields or changing the existing baseline-selection policy.
- [ ] External adapters consume the per-iteration tensors supplied by the
      benchmark protocol, with any deliberately static plan metadata declared.
- [ ] FP8 GQA uses the suite-default warmup/repeat protocol.
- [ ] Focused correctness and benchmark-protocol tests cover every changed
      family.

**Comment by superAngGao** [comment]
## Class 1: timed baseline implementations that need correction

Implementation is tracked in #1881.

These are not merely labeling problems: the timed callable itself contains
work that should not be part of the competitive framework baseline.

| Operators | Current problem | Concrete correction |
| --- | --- | --- |
| `SumFwdOp`, `MeanFwdOp`, `ProdFwdOp` | Explicit `x.float()`, reduction, and output cast | Call `torch.sum`, `torch.mean`, or `torch.prod` directly on the contract input dtype. Let PyTorch choose its internal accumulation. |
| `StdFwdOp`, `VarFwdOp` | Explicit full-input FP32 materialization | Call `torch.std` / `torch.var` directly with the manifest dimensions and correction. |
| `VarMeanFwdOp` | Runs separate `var` and `mean`, including two FP32 conversions | Use one direct `torch.var_mean` call. |
| `L1NormFwdOp`, `L2NormFwdOp`, `InfNormFwdOp` | Materializes an FP32 copy before `torch.linalg.vector_norm` | Call `torch.linalg.vector_norm` directly on the original tensor with the matching `ord` and dimensions. |
| `CumsumFwdOp`, `CumprodFwdOp` | Times FP32 input/output conversions | Use direct `torch.cumsum` / `torch.cumprod` with the contract dtype. |
| `RMSNormFwdOp` | Manually decomposes RMSNorm in FP32 even though the manifest declares `F.rms_norm` | Add direct `F.rms_norm` as the framework baseline; retain the decomposition only as the numerical oracle. Also add FlashInfer under Class 2. |
| `BatchNormFwdOp` | Times input conversion and clones both running-stat tensors; output dtype does not match the contract | Call native `F.batch_norm` on FP16/BF16 shifting-pool inputs and their per-iteration mutable running stats. Do not clone inside the timed callable. |
| `BatchNormBwdOp` | Rebuilds and executes an FP32 autograd graph despite receiving saved mean/rstd | Call `aten.native_batch_norm_backward` with the supplied saved statistics. Keep autograd as reference-only. |
| `FusedAddLayerNormFwdOp` | Correctness path converts add result, weight, and bias to FP32 | Keep this path as the oracle. Until an exact external fused kernel is available, use same-dtype `add + F.layer_norm` as the eager performance baseline and an exact compiled version as the optimized framework baseline. |
| `FusedAddRMSNormFwdOp` | Correctness path materializes FP32 intermediates | Keep it as the oracle; use `flashinfer.norm.fused_add_rmsnorm` as the performance baseline (Class 2). |

Representative H200 A/B already measured on the audited commit:

| Workload | Current baseline | Correct direct API | Inflation |
| --- | ---: | ---: | ---: |
| sum | 0.046831 ms | 0.011968 ms | 3.913x |
| mean | 0.046784 ms | 0.011968 ms | 3.909x |
| var_mean | 0.110400 ms | 0.034736 ms | 3.178x |
| L1 norm | 0.046976 ms | 0.017488 ms | 2.686x |
| RMSNorm prefill | 0.142367 ms | 0.012561 ms | 11.335x |
| RMSNorm decode | 0.061264 ms | 0.007425 ms | 8.252x |

BatchNorm needs the same treatment: the representative forward baseline drops
from about 0.0385 ms to 0.0056 ms, and backward from about 0.1232 ms to
0.0119 ms when the native contract-matching paths are used.

**Comment by superAngGao** [comment]
## Class 2: replace reference-only timing with real performance baselines

The existing reference remains the correctness oracle. The table specifies the
additional implementation that should provide the performance comparison.

### Exact or close production-library coverage

| Operators | Proposed performance baseline | Required adapter work |
| --- | --- | --- |
| `RMSNormFwdOp` | `flashinfer.norm.rmsnorm` | Match epsilon/output dtype. |
| `FusedAddRMSNormFwdOp` | `flashinfer.norm.fused_add_rmsnorm` | FlashInfer mutates input/residual in place; return them as `(output, residual_out)` and rely on the shifting pool for fresh iteration inputs. |
| `SiluAndMulFwdOp`, `GeluAndMulFwdOp`, `GeluTanhAndMulFwdOp` | `flashinfer.activation.{silu,gelu,gelu_tanh}_and_mul` | Match packed `[gate, value]` layout and exact/tanh GELU mode. |
| `RopeNeoxOp`, `RopeNonNeoxOp`, position-ID variants | FlashInfer `apply_rope*` APIs | Build the cos/sin cache before timing; pass the per-iteration data tensors. Select `is_neox` correctly. |
| `RopeLlama31Op` | FlashInfer `apply_llama31_rope*` | Map Llama-3.1 scaling parameters exactly. |
| `RopeYarnOp` | vLLM `yarn_scaling_rope` | Instantiate/cache the rotary object before timing. |
| `RopeLongRopeOp` | vLLM `phi3_long_rope_scaled_rope` | Map short/long rescale factors and position limits. |
| `MHCPreOp`, `MHCPostOp` | FlashInfer `mhc_pre_big_fuse*` / `mhc_post`, with Liger MHC as a second candidate | Upgrade FlashInfer: installed `0.6.11.post2` lacks `flashinfer.mhc`; current FlashInfer exposes these APIs. Validate HC/layout and mixed BF16/FP32 inputs. |
| Gated Delta / Delta decode operators | FlashInfer `gdn_decode.*` and/or FLA | Select the implementation whose state orientation and returned updated state match the manifest; do not time transposes that TileOps does not require without labeling the comparison. |
| `SSDDecodeOp` | FlashInfer `mamba.selective_state_update` or Mamba-SSM selective-state update | Validate the in-place FP32 state layout and returned `y_out`. |
| `MoeGroupedGemmNopadFwdOp` | `flashinfer.grouped_mm.grouped_mm_bf16` | Convert `true_sizes/true_offsets` to `m_indptr` outside timing only if they are declared plan metadata; otherwise include conversion. |
| `FusedMoeFwdCbFwdOp`, `FusedMoEExpertsNopadPersistent3WGFwdOp` | FlashInfer CUTLASS fused MoE and/or vLLM fused MoE | Match routing weights/IDs, correction bias, activation, deterministic reduction, weight layout, and output dtype. |
| `FP8QuantOp` | vLLM `QuantFP8` dynamic per-token/group path | Flatten `[B,S,G,D]` to rows, quantize per row, and reshape FP32 scales to `[B,S,G]`. Verify scale convention before accepting latency. |
| `GroupedQueryAttentionPrefillFwdOp`, `GroupedQueryAttentionPrefillVarlenFwdOp` | FA3 and FlashInfer prefill | Pass causal/window/softcap/scale parameters; keep manual attention only as an oracle. |
| `GroupedQueryAttentionPrefillPagedWithKVCacheFwdOp` | FlashInfer paged prefill | Tracked in #1837, including FP8-cache workloads. |
| `MultiHeadLatentAttentionDecodeWithKVCacheFwdOp` | FlashInfer MLA decode | Plan outside timing, run on the per-iteration Q/KV tensors. |
| `DeepSeekSparseAttentionDecodeWithKVCacheFwdOp`, `FP8LightningIndexerOp` | vLLM DSA/lightning-indexer implementation | Candidate is present in current vLLM; first prove exact index range, scale, cleanup, FP8 scale, and output-layout compatibility. |

The current H200 environment already has FlashInfer `0.6.11.post2`, vLLM
`0.19.1`, FLA `0.5.2`, and Mamba-SSM `2.3.1`. RMSNorm,
FusedAddRMSNorm, gated activations, standard RoPE, GDN decode, and grouped MM
are therefore available without adding a new dependency. Only the newer
FlashInfer MHC surface requires an upgrade. Liger is an optional independent
baseline, not a prerequisite for the first patch.

### No exact general third-party contract found

| Operators | Performance fallback | Reason not to force a third-party adapter |
| --- | --- | --- |
| `FusedAddLayerNormFwdOp` | Same-dtype `add + F.layer_norm`, plus `torch.compile(fullgraph=True)` | FlashInfer's DiT fused residual LayerNorm kernels are shape/formula restricted and do not cover the general weight+bias contract. |
| `AdaLayerNormFwdOp`, `AdaLayerNormZeroFwdOp` | Exact same-dtype PyTorch expression, compiled as one graph | Available DiT kernels commonly implement `(1 + scale)` and restricted layouts, while TileOps declares `scale * LayerNorm(x) + shift` and an optional outer gate. |
| `EngramGateConvFwdOp`, `EngramGateConvBwdOp`, `EngramDecodeOp` | Exact benchmark-local graph compiled ahead of timing | No contract-equivalent public production kernel was identified. The autograd/high-precision path remains the oracle. |
| `AlibiFwdOp`, `SinusoidalFwdOp` | Exact same-dtype compiled PyTorch graph | Attention libraries consume these effects inside attention rather than exposing the standalone tensor-producing contract. |
| `MeanPoolingForwardOp` | Direct same-dtype `torch.mean`; optionally compiled | A native framework primitive is the appropriate performance baseline. |
| `AdaptiveMaxPool2dFwdOp`, `AdaptiveMaxPool2dIndicesFwdOp`, `MaxPool{1,2,3}FwdOp`, index variants, `AvgPool{1,2,3}FwdOp`, `AdaptiveAvgPool2dFwdOp` | Direct `torch.nn.functional` pooling API | These are cuDNN/native framework performance paths, not merely correctness references. |

If the candidate external API cannot reproduce the manifest contract without
changing semantics, it must not be used. An exact compiled graph can be added
when it is useful and stable; otherwise the accurately tagged high-precision
reference remains the only baseline. Do not drop the comparison merely because
no production-library equivalent exists.

**Comment by superAngGao** [comment]
## Class 3: fix misleading tags using the existing report schema

### Confirmed operator-level mislabeling

| Operator | Current report | Actual implementation | Correction |
| --- | --- | --- | --- |
| `EngramGateConvBwdOp` | `torch` | Benchmark-local custom/autograd reference | Rename the current result to `torch-autograd-ref`; use the Class-2 performance path separately. |
| `DeltaNetDecodeOp` | `torch` | Multi-op pure-PyTorch decomposition | Rename it to `torch-ref`; add `fla` or `flashinfer` when the state contract matches. |
| `GroupedQueryAttentionDecodeWithKVCacheFwdOp` softcap case | `torch-sdpa` and 56.3x | Manual `repeat_interleave + FP32 matmul + softmax + matmul` fallback | Pass `softcap`/`logits_soft_cap` and `sm_scale` into FA3/FlashInfer; rename the manual fallback to `torch-fp32-ref`. |
| `GemmFp8Op` | `torch-scaled-mm` | Expands scales, converts both operands to FP32, then runs FP32 `torch.matmul` | Rename it to `torch-fp32-ref`. Add actual `torch._scaled_mm` under `torch-scaled-mm` for compatible per-tensor cases and retain FlashInfer for block-scale performance. |

The FP8 GEMM issue is especially concrete: the current callable never invokes
`torch._scaled_mm`, although the installed Torch 2.10 exposes the native
`aten::_scaled_mm` schema.

### Tag convention

Do not change the manifest or add report fields. Keep the existing
`BenchmarkReport.record(..., tag=...)` interface and use an explicit naming
convention:

```python
BenchmarkReport.record(
    ...,
    tag="flashinfer",
)

BenchmarkReport.record(
    ...,
    tag="torch-fp32-ref",
)
```

The convention is descriptive, not a new filtering rule:

- `torch-ref`, `torch-fp32-ref`, and `torch-autograd-ref` accurately tell the
  reader that the number comes from a benchmark-local reference;
- `torch`, `torch-scaled-mm`, `flashinfer`, `fa3`, `fla`, and similar tags must
  only be used when that implementation was actually timed;
- a reference may remain the sole baseline when no exact external or native
  implementation exists;
- adding a stronger baseline supplements the reference rather than deleting
  it;
- the dashboard keeps its existing display, badge, regression, and selection
  behavior. This issue does not infer semantics from a suffix.

The serialized format stays unchanged:

```text
BenchmarkReport.record(tag=...)
  -> existing pytest/JUnit tag properties
  -> existing nightly JSON/Markdown tag fields
  -> dashboard/history
```

Add focused tests that assert the corrected operator paths emit the expected
existing tag and that no callable is recorded under an implementation name it
does not invoke. No generic tag classifier or report filter is introduced.

**Comment by superAngGao** [comment]
## Class 4: per-op deviations from the suite benchmark protocol

### A. External baselines that bypass the shifting input pool

`bench_kernel(fn, args=...)` rotates tensor arguments per iteration. A
zero-argument closure receives `input_policy=none`, and a callable that ignores
its arguments can silently keep using the setup-time tensor even when
`bm.profile(..., *inputs)` appears correct.

| Affected operators/adapters | Current behavior | Correction |
| --- | --- | --- |
| `DeltaNetFwdOp`, `DeltaNetBwdOp`, combined `DeltaNetOp` FLA paths | Zero-argument closures capture one FLA-layout tensor set | Prebuild an FLA-layout shifting pool and pass each iteration's tensors as callable arguments. For backward, build equivalent per-sample graph state or explicitly document/validate the retained-graph policy. |
| GatedDeltaNet forward/backward/combined FLA paths | Same zero-argument closure pattern | Same native-layout shifting-pool adapter. |
| GLA forward/backward and GLADecode FLA paths | Same zero-argument closure pattern | Same native-layout shifting-pool adapter. |
| GatedDeltaNetDecode FLA path | Captures fixed decode inputs/state | Pass per-iteration data/state explicitly and preserve required in-place state semantics. |
| `DaCumsumFwdOp`, `SSDChunkScanFwdOp`, `SSDChunkStateFwdOp`, `SSDStatePassingFwdOp` Mamba paths | Zero-argument closures capture original tensors | Pass Mamba-layout tensors as benchmark arguments. `.contiguous()` must be performed during pool construction, not hidden in the timed invocation. |
| `MultiHeadAttentionDecodePagedWithKVCacheFwdOp` FA3/FlashInfer | Adapter captures setup-time K/V pages; callable parameters are ignored | Reshape the K/V tensors supplied for the current iteration and pass them to FA3/FlashInfer `run`. Keep reusable workspace/plan outside timing. |
| `GroupedQueryAttentionDecodePagedWithKVCacheFwdOp` FA3/FlashInfer | Same captured paged-KV pattern | Same correction, including per-iteration Q and KV data. |
| Non-paged `GroupedQueryAttentionDecodeWithKVCacheFwdOp` FlashInfer | Single-request path captures Q/K/V; batched path captures K/V | Use the arguments supplied to each invocation; wrapper plan/workspace may remain static. |

Static planning metadata may remain fixed only when it is not mutated by the op
and its fixed-pointer use matches the intended serving contract. Record that
choice explicitly, for example `input_policy="shifting-data+static-plan"`.

### B. FP8 GQA sampling override

`bench_gqa_fp8.py` explicitly runs both TileOps and FA3 with:

```python
n_warmup=1, n_repeat=3, n_trials=1
```

The suite protocol is 10 warmups and 50 timed samples with median aggregation.
Remove the override and use the defaults. If a reduced smoke mode is still
needed, give it a separate marker/protocol identifier and do not merge its
numbers into the nightly performance table.

### Verification

- Assert TileOps and every non-reference baseline result expose compatible
  `input_policy`, `n_warmup`, `n_repeat`, `n_trials`, activity kinds, and timing
  protocol metadata.
- Add synthetic/unit coverage that rejects a performance baseline whose
  callable declares tensor arguments but never consumes the rotating data.
- Run focused A/B for representative multi-kernel FLA/Mamba and paged-attention
  cases after adapter repair; correctness must be checked on more than the
  setup-time tensor set.
